src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs

```diff
@@ -520,21 +528,22 @@ internal abstract partial class SafeDeleteContext : SafeHandle
 
                                 if (outUnmanagedBuffer.cbBuffer > 0)
                                 {
-                                    if (outSecBuffer.size == 0)
-                                    {
-                                        // We did not get anything in the first round.
-                                        outSecBuffer.size = outUnmanagedBuffer.cbBuffer;
-                                        outSecBuffer.type = outUnmanagedBuffer.BufferType;
-                                        outSecBuffer.token = new Span<byte>((byte*)outUnmanagedBuffer.pvBuffer, outUnmanagedBuffer.cbBuffer).ToArray();
-                                    }
-                                    else
+                                    int needed = outSecBuffer.size + outUnmanagedBuffer.cbBuffer;
+                                    if (outSecBuffer.token == null || outSecBuffer.token!.Length < needed)
                                     {
-                                        byte[] buffer = new byte[outSecBuffer.size + outUnmanagedBuffer.cbBuffer];
-                                        Buffer.BlockCopy(outSecBuffer.token!, 0, buffer, 0, outSecBuffer.size);
-                                        new Span<byte>((byte*)outUnmanagedBuffer.pvBuffer, outUnmanagedBuffer.cbBuffer).CopyTo(new Span<byte>(buffer, outSecBuffer.size, outUnmanagedBuffer.cbBuffer));
-                                        outSecBuffer.size = buffer.Length;
+                                        byte[] buffer = ArrayPool<byte>.Shared.Rent(outSecBuffer.size);
```

Comment:
That echoes sentiment from @rzikm ... but I'm not sure how to address it.
The general idea is that the PAL layer will rent the array and everything is returned up high in `finally` of `ForceAuthenticationAsync`. That is core function of the TLS handshake (and renegotiation). Because we have multiple PALs you will see more renting than returning.  

BTW the section in SspiCli is particularly complicated as we possibly need to merge data from two calls e.g. we may need to get bigger buffer and copy the data. While it happened in field we do not have clear test case that would exercise this path. 

In the past, we did not drag `ProtocolToken` to PAL layer - but we could if that makes it more readable. Perhaps via `ref` and we can add finalizer to make sure it is always released. 

Since this already missed 8.0 we have more time for alternatives. What do you think about wrapping native memory in `Memory<byte>` @stephentoub? I would allow us to avid the array manipulations completely.